src/cargo/core/package_id_spec.rs

```diff
@@ -101,16 +105,57 @@ impl PackageIdSpec {
             name: String::from(package_id.name().as_str()),
             version: Some(package_id.version().clone().into()),
             url: Some(package_id.source_id().url().clone()),
+            kind: Some(package_id.source_id().kind().clone()),
         }
     }
 
     /// Tries to convert a valid `Url` to a `PackageIdSpec`.
     fn from_url(mut url: Url) -> CargoResult<PackageIdSpec> {
-        if url.query().is_some() {
-            bail!("cannot have a query string in a pkgid: {}", url)
+        let mut kind = None;
+        if let Some((kind_str, scheme)) = url.scheme().split_once('+') {
+            match kind_str {
+                "git" => {
+                    let git_ref = GitReference::from_query(url.query_pairs());
```

Comment:
> git ref query strings for URLs

We've got a fair amount of issue about git dependencies ambiguity because of not checking in the actual rev. This patch seems to fix <https://github.com/rust-lang/cargo/issues/10256>. Do you want to include that demo test case as well?